data/transactions/logic/eval.go

```diff
@@ -612,26 +612,226 @@ func (cx *EvalContext) RunMode() RunMode {
+	name := at.String()
+	if len(stname) > 0 {
+		name = stname[0]
+	}
+
+	st := StackType{Name: name, AVMType: at}
+
+	switch at {
+	case avmBytes:
+		st.LengthBound = bounds
+	case avmUint64:
+		st.ValueBound = bounds
+	}
+
+	return st
+}
+
+func (st StackType) narrowed(bounds [2]uint64) StackType {
+	// It's static, set the name to show
+	// the static value
+	if bounds[0] == bounds[1] {
+		switch st.AVMType {
+		case avmBytes:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("[%d]byte", bounds[0]))
+		case avmUint64:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("%d", bounds[0]))
+		}
+	}
+	return NewStackType(st.AVMType, bounds)
+}
+
+// AssignableTo returns a bool indicating whether the receiver can be
+// assigned to some other type that is expected by the next operation
+func (st StackType) AssignableTo(other StackType) bool {
+	if st.AVMType == avmNone || other.AVMType == avmNone {
+		return false
+	}
+
+	if st.AVMType == avmAny || other.AVMType == avmAny {
+		return true
+	}
+
+	// By now, both are either uint or bytes
+	// and must match
+	if st.AVMType != other.AVMType {
+		return false
+	}
+
+	// Same type now
+
+	// Check if our constraints will be satisfied by
+	// the other type
+	switch st.AVMType {
+	case avmBytes:
+		smin, smax := st.LengthBound[0], st.LengthBound[1]
+		omin, omax := other.LengthBound[0], other.LengthBound[1]
+
+		// yes definitely
+		// [32,32] => [0..4k]
+		// [32,32] => [32,32]
+
+		// yes, maybe determined at runtime
+		// [0..4k] => [32,32]
+
+		// no, cant fit
+		// [64,64] => [32,32]
+		// no, makes no sense
+		// [32,32] =>  [64,64]
+
+		// we only have 0-N and [N,N] (static) and only
+		// those that are both not static and have different lengths
+		// can be assigned
+		return !(smin == smax && omin == omax && smin != omin)
+
+	case avmUint64:
+		// No static values at compile
+		// time so hard to do any typechecks for assembler,
+		// dont use this for avm runtime
+		return true
+	default:
+		panic("no stack type match in AssignableTo check")
+	}
+}
+
+func (st StackType) String() string {
+	return st.Name
+}
+
+// Typed tells whether the StackType is a specific concrete type.
+func (st StackType) Typed() bool {
+	switch st.AVMType {
+	case avmUint64, avmBytes:
+		return true
+	}
+	return false
+}
+
 // StackTypes is an alias for a list of StackType with syntactic sugar
 type StackTypes []StackType
 
+// Reverse returns the StackTypes in reverse order
+// useful for displaying the stack as an op sees it
+func (st StackTypes) Reverse() StackTypes {
+	nst := make(StackTypes, len(st))
+	for idx := 0; idx < len(st); idx++ {
+		nst[idx] = st[len(st)-1-idx]
+	}
+	return nst
+}
+
+func (st StackTypes) String() string {
+	// Note this reverses the stack so top appears first
+	return fmt.Sprintf("(%s)", strings.Join(st.strings(), ", "))
+}
+
+func (st StackTypes) strings() []string {
+	var strs = make([]string, len(st))
+	for idx, s := range st {
+		strs[idx] = s.String()
+	}
+	return strs
+}
```

Comment:
with generics, becomes the same code as `typeStrings()`

Reply:
I ended up changing `typeStrings` to omit `"none"` entries